sdk/src/main/java/io/customer/sdk/util/JsonAdapter.kt

```diff
@@ -0,0 +1,74 @@
+package io.customer.sdk.util
+
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.Types
+
+/**
+ * Abstract way to deserialize JSON strings without thinking about the library used.
+ *
+ * This is an object instead of a class through dependency injection because when we run tests, we like to use the real instance of this class in our tests. That is usually a sign it doesn't need to be injected.
+ */
+class JsonAdapter internal constructor(val moshi: Moshi) {
+
+    /**
+     * Note: This class must treat arrays and not arrays differently (below we call them Lists but we mean arrays because that's what we call them in Json).
+     *
+     * We are not talking about embedded arrays:
+     * ```
+     * {
+     *   "nested": []
+     * }
+     * ```
+     * Nesting works fine with the regular functions.
+     *
+     * We mean:
+     * ```
+     * {
+     *   "not_array_json": ""
+     * }
+     * ```
+     * vs
+     * ```
+     * [
+     *   {
+     *     "not_array_json": ""
+     *   }
+     * ]
+     * ```
+     *
+     * Moshi handles arrays differently: https://github.com/square/moshi#parse-json-arrays
+     */
+    inline fun <reified T : Any> fromJson(json: String): T {
+        val json = json.trim()
+
+        if (json.isNotEmpty() && json[0] == '[') throw IllegalArgumentException("String is a list. Use `fromJsonList` instead.")
+
+        val jsonAdapter = moshi.adapter(T::class.java)
+
+        return jsonAdapter.fromJson(json) as T
+    }
+
+    inline fun <reified T : Any> fromJsonList(json: String): List<T> {
+        val json = json.trim()
+
+        if (json.isNotEmpty() && json[0] != '[') throw IllegalArgumentException("String is not a list. Use `fromJson` instead.")
```

Comment:
do we need to trim it, if we are to check the 1st character in a string?

Reply:
Are you asking a question about `val json = json.trim()` or `if (json.isNotEmpty() && json[0] != '[') throw IllegalArgumentException("String is not a list. Use `fromJson` instead.")`? The comment is made on line 54 but I think you're asking a question about line 52?

Reply:
Sorry, missed the `trim` on line 52, that's why I was suggesting to have it in like 54.